### DISCUSSION TITLE
function returned from expression gets called in x-init

### DISCUSSION THREAD

**Discussion by tvdr** [alpinejs/alpine#3695] [1. Help]
In Filamentphp, some elements utilize x-init="Livewire.hook(...)"  to define a commit hook for the component.
Livewire.hook returns an "off" function (according to the comments in the code) for disabling the listener.

with x-init="Livewire.hook(...)" the alpine evaluator invokes the returned function, thus the hook immediately deregister.

I was wondering if this is intentional or would be considered as a bug?

[Codepen with some examples](https://codepen.io/tvdr/pen/JjezvNN?editors=1111)

**Comment by ekwoka** [comment]
So, the way the expression evaluator works, is that it will keep running any functions returned by the expression until its not a function anymore.

This is because expressions can point to functions, or call them in the expression, and there can be multiple layers of this that happen to handle allowing the edge cases to do everything that people expect the expression to do.

Exactly what is the best way to do what you're trying to do would be hard for me to recommend, since I don't know enough about livewire, but I suspect this pattern of calling the hooks in an x-init is not the intended use of them.

If you make your expression an invalid Right hand expression, then it will be wrapped and isolated. So like `const cleanup = Livewire.hook(); destroy = cleanup` This would not immediately invoke it, and would store it on the component in the `destroy` key, which should also be invoked on component cleanup. Though make sure you assign the key `destroy` on the immediate parent.
